### ISSUE TITLE
Multiple subtitle functionality and some ui functionality

### ISSUE THREAD

**Issue by BhaturaGuy** [mpv-android/mpv-android#495]
External fonts can't be used on mpv-andriod as of now but almost every other media player does support that with extra features which are good(subjective) in their own way.

### Additional information

Addition of multiple extra settings like external font use for "srt" subs.

For example: 
![Screenshot_2022-02-05-21-13-54-01_f2cb81fb7cf38af7978f186f2a61634a__01](https://user-images.githubusercontent.com/93109835/153427506-7a262024-3ba5-4f33-8ef1-1672a699b5d8.jpg)

This is a screenshot of Mx Player where there are a lot of configs for subtitles, some of those settings are configurable through mpv.conf but then again, people don't usually like typing all of that or studying what to add or not.

Like this fade out option looks good, it fades every subtitle which has a gap of around 2 secs between that and next to that subtitle. Don't know if that's possible on mpv on PCs.

Now for some ui feature requests:
![Screenshot_2022-02-10-20-01-54-79_f2cb81fb7cf38af7978f186f2a61634a](https://user-images.githubusercontent.com/93109835/153429059-9bd26d9f-cde3-4df9-9a3f-949ee22f9ddc.jpg)
![Screenshot_2022-02-10-20-01-05-62_4aed3257f278fcf7bfa3abd644e23333](https://user-images.githubusercontent.com/93109835/153429072-6f45cb1b-1db4-4689-87ae-86ae82e64c53.jpg)

Now, these are 2 screenshots of Mx Player and VLC playing a video. Notice how these two have multiple floating buttons for different settings according to average user's requirement, since mpv doesn't have that functionality, there are only some keybinds which are not very helpfull in my opinion, it removes tap to hide, not exactly much keybinds to use or actually possible to use, floating buttons are really helpful. One way could be that set them and the amount of them as configurable. Mpv aspect-ratio settings also aren't that reliable unlike vlc or mx player's, Fit to screen option is not available in mpv and modern phones have a different aspect ratio nowadays and mpv's 21:9 aspect ratio isn't actually helpful cause nobody wants to see stretched characters where the top and bottom still have black bars.

Tl;dr New ui settings and subtitle configuration that are _meant_ to be on an _android_ device.

**Comment by sfan5** [comment]
Thanks for your feedback, most people just say the UI is bad (which I can understand) without being exact about what they would like changed.

> This is a screenshot of Mx Player where there are a lot of configs for subtitles, some of those settings are configurable through mpv.conf but then again, people don't usually like typing all of that or studying what to add or not.

It's worth noting that mpv is a more technical player than most. I agree that UI would be nice but configuring these via mpv.conf is *possible* today so that's kind of lower priority.

> Mpv aspect-ratio settings also aren't that reliable unlike vlc or mx player's, Fit to screen option is not available in mpv

Fit to screen is possible, it's just "hidden" in the gesture menu. The option is called Zoom (internally this is called [panscan](https://mpv.io/manual/master/#options-panscan)). Adding it into the same menu with aspect ratios would make sense.

**Comment by qgustavor** [comment]
One thing I want to add is that mpv supports multiple subtitles but implements differently from MX Player: seems MX just overlays multiple renderers one above the other, resulting in overlapping subtitles, mpv removes styling from the secondary subtitle and moves it to the top part of the screen. mpv's implementation is good if you want to watch something translated to two languages to learn one of those, MX's is good if you have a main subtitle and a secondary subtitle with extra parts and want to watch both together.

Because I often find the MX's implementation more useful to me (i.e. many videos I watch have subtitles with just dialogue and subtitles with just songs/signs and I want to watch both) I modified mpv code (the Windows build) to be like MX [by commenting this line](https://github.com/mpv-player/mpv/blob/9ffaa6b81bb78d9813b2f10ea56e51b700660e2d/player/sub.c#L183). Looks it's hard to make it toggleable via a config: [this issue](https://github.com/mpv-player/mpv/issues/3022) was been open for a lot of time.

Anyway, that's not important: while is possible to set up a lot of things in mpv.conf there are not a lot of inputs to bind on. Creating a GUI for every single mpv.config is impossible: there are a lot of possibilities within mpv! One example: some people like to change subtitle scale [using add](https://github.com/mpv-player/mpv/blob/9ffaa6b81bb78d9813b2f10ea56e51b700660e2d/etc/input.conf#L101-L102), in MX you can pinch zoom, but in desktop mpv I prefer using `cycle-values sub-scale 1 1.2 1.5`.

I hope someone find a good idea to improve the UX and UI of this project. It's hard! Maybe, instead of just comparing it to completely different players, how about comparing it to mpv.net? It adds a GUI to edit many of the most used mpv.conf settings, including default subtitle styling, but not to an extreme (like it don't adds anything relative to secondary subtitles) but those features can still be used by advanced users by editing config files.

Some may look at this problem thinking "what are the functions that 90% of the users would use in a player?", but how to get this answer? Asking people in GitHub may help, but people here are not the majority. I'm not part of the majority of the users: sure most people would not even think to compile mpv themselves. Using some kind of analytics is not good: just see how Audacity users reacted. It's hard.

There a lot of open issues related to UI, but I will comment about this just once to avoid spam and, because this issue talked about MX, I wrote all of the above in this issue. I wish I could help developing, I will try to check the code and implement something to improve this issue, but I'm really inexperienced at this. Good luck people.

**Comment by qgustavor** [comment]
@Valllent Comment this line before building mpv: https://github.com/mpv-player/mpv/blob/9ffaa6b81bb78d9813b2f10ea56e51b700660e2d/player/sub.c#L183

I wish I knew how to make it to be a config in mpv.conf then propose it via a pull request. I still have a lot to learn…

**Comment by ValentinPostindustria** [comment]
@qgustavor Sorry for bothering you. But it didn't help.  I still can choose only one subtitle language. Maybe I skipped something? Or there are some actions that I should do to display the second subtitles on the top? Because the UI allows choosing only one per time.  

I tried more than 50 video players from Play Store before, but there is no player that can show double subtitles in the separated positions. And accidentally I found your comment, so you are my last chance to learn another language by simply watching movies on Android.

**Comment by qgustavor** [comment]
If you are talking about two with the second above then you don't need to change mpv's code, that's the default behavior. I use secondary subtitles for songs, signs and other things that people put in secondary tracks so I thought you wanted the same after seeing my comments.

At the moment the main distribution of mpv-android do not have a GUI to choose secondary subtitles but you can edit input.conf and bind one of supported inputs to `cycle secondary-sid`. That is what I did before forking android-mpv.

In my fork I changed the subtitle button behavior: tapping open the main subtitle selector interface, holding it open the secondary subtitle interface. It looks like this: https://imgur.com/a/I11kU0G (track 1 contains song translations, track 2 is BD subtitles, 3 is HDTV subtitles). If you want you can check [my code](/qgustavor/mpv-android/), I published it. I will not provide apks, instead I hope the features I implemented can be reproduced here in the future when better scripting support gets implemented (which is in the plans IIRC).

**Comment by ValentinPostindustria** [comment]
@qgustavor, you are the best. Thank you. I tried your fork without commenting `sub.c` file. Also, I added a null pointer checker on the 1102 line of `MPVActivity` to fix exceptions on my device when clicking "Settings".  And it works as needed. Now I can choose 2 subtitles and apply any delay to them like on my PC. 

I hope that in future versions of MPV it will be possible to choose 2 subtitles using UI interface.
